### PR TITLE
ADEN-8996 Adding Viewability Counter service

### PR DESCRIPTION
**Pull request by dmnsobczak**

### CODE REVIEW

File: src/ad-services/viewability-counter/index.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import { context, Dictionary, LocalStorage, utils } from '@ad-engine/core';
```
Comment (fraszczakszymon) [inline src/ad-services/viewability-counter/index.ts]: I believe `ad-services` package is desired for external integrations. What about putting it in the `src/ad-engine/services`?

File: examples/context.ts [hunk 2]
```diff
@@ -376,6 +376,10 @@ export default {
 			enabled: true,
 			partnerCode: 'wikiaprebidheader490634422386',
 		},
+		viewabilityCounter: {
```
Comment (fraszczakszymon) [inline examples/context.ts]: Same here. It's not external service

File: src/ad-engine/services/local-storage.ts [hunk 3]
```diff
@@ -3,11 +3,13 @@ import { Dictionary } from '../models';
 export class LocalStorage {
 	private fallbackStorage: Dictionary = {};
 
+	constructor(private storage: any = window.localStorage) {}
```
Comment (fraszczakszymon) [inline src/ad-engine/services/local-storage.ts]: It is going to fail when localStorage is not available. We have `localStorage` module in the `Wikia/ad-engine`, would it be useful for that case?

File: examples/context.ts (hunk 2 above)
Comment (dmnsobczak) [reply examples/context.ts]: Moved to `options`

File: src/ad-services/viewability-counter/index.ts (hunk 1 above)
Comment (dmnsobczak) [reply src/ad-services/viewability-counter/index.ts]: Oh, that's a cool place. Moved.

**Comment by dmnsobczak** [comment]
Changed to use cookieStorage.

File: src/ad-engine/services/viewability-counter.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import { Dictionary } from '../models';
+import { logger } from '../utils';
+import { context } from './context-service';
+import { LocalStorage } from './local-storage';
+import { sessionCookie } from './session-cookie';
+
+type ViewabilityStatus = 'loaded' | 'viewed';
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]: ViewabilityStatus -> StatusType?

File: src/ad-engine/services/viewability-counter.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		this.counters = this.storage.getItem('viewabilityCountData') || {
+			loadedCounter: {},
+			viewedCounter: {},
+		};
+	}
+
+	/**
+	 * Save in storage impression or viewability change
+	 */
+	updateStatus(type: ViewabilityStatus, slotName: string): void {
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]: updateStatus -> incrementStatusCounter?

File: src/ad-engine/services/viewability-counter.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import { logger } from '../utils';
+import { context } from './context-service';
+import { LocalStorage } from './local-storage';
+import { sessionCookie } from './session-cookie';
+
+type ViewabilityStatus = 'loaded' | 'viewed';
+
+const logGroup = 'viewability-service';
+
+/**
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]: Aren't we removing jsdocs? :P (in favor of TS)

File: src/ad-engine/services/viewability-counter.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import { Dictionary } from '../models';
+import { logger } from '../utils';
+import { context } from './context-service';
+import { LocalStorage } from './local-storage';
+import { sessionCookie } from './session-cookie';
+
+type ViewabilityStatus = 'loaded' | 'viewed';
+
+const logGroup = 'viewability-service';
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]:
```suggestion
const logGroup = 'viewability-counter';
```

File: src/ad-engine/utils/geo.ts [hunk 8]
```diff
@@ -109,19 +91,7 @@ function synchronizeCookie(): void {
 		}
 	});
 
-	setCookie(JSON.stringify(cachedVariables));
-}
-
-function setCookie(value: any): void {
```
Comment (fraszczakszymon) [inline src/ad-engine/utils/geo.ts]: 💎

File: src/ad-engine/services/session-cookie.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	removeItem(key: string): void {
+		Cookies.remove(`${this.prefix}_${key}`, {
+			path: '/',
+			domain: this.getCookieDomain(),
+		});
+	}
+
+	clear(): void {
```
Comment (fraszczakszymon) [inline src/ad-engine/services/session-cookie.ts]: Seems not used?

File: src/ad-engine/services/session-cookie.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		if (!this.keysSeen.includes(key)) {
+			this.keysSeen.push(key);
+		}
+
+		Cookies.set(`${this.prefix}_${key}`, input, cookieAttributes);
+
+		return true;
+	}
+
+	removeItem(key: string): void {
```
Comment (fraszczakszymon) [inline src/ad-engine/services/session-cookie.ts]: Seems not used?

File: src/ad-engine/services/viewability-counter.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+			!context.get('options.viewabilityCounter.enabled') ||
+			(context.get('options.viewabilityCounter.ignoredSlots') &&
+				context.get('options.viewabilityCounter.ignoredSlots').includes(slotName))
+		) {
+			return;
+		}
+
+		logger(logGroup, 'storing viewability status', type, slotName);
+
+		this.counters[`${type}Counter`][slotName] =
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]: slotName / viewabilityCounterId then

File: src/ad-engine/services/session-cookie.ts (hunk 10 above)
Comment (dmnsobczak) [reply src/ad-engine/services/session-cookie.ts]: This class is used as a dependency for Local Storage. I want to keep the same interface for all storages.

File: src/ad-engine/services/session-cookie.ts (hunk 9 above)
Comment (dmnsobczak) [reply src/ad-engine/services/session-cookie.ts]: This class is used as a dependency for Local Storage. I want to keep the same interface for all storages.

File: src/ad-engine/services/viewability-counter.ts (hunk 4 above)
Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Done!

File: src/ad-engine/services/viewability-counter.ts (hunk 7 above)
Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Done!

File: src/ad-engine/services/viewability-counter.ts (hunk 6 above)
Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Ohh fine. But I like those comments :D

File: src/ad-engine/services/viewability-counter.ts (hunk 5 above)
Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Ok, why not. Changed!

File: src/ad-engine/utils/geo.ts (hunk 8 above)
Comment (dmnsobczak) [reply src/ad-engine/utils/geo.ts]: 👍

File: src/ad-engine/services/viewability-counter.ts (hunk 11 above)
Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Renamed!

File: src/ad-engine/services/viewability-counter.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+				this.incrementStatusCounter('loaded', id);
+			});
+
+			slot.viewed.then(() => {
+				this.incrementStatusCounter('viewed', id);
+			});
+		});
+	}
+
+	incrementStatusCounter(type: StatusType, counterId: string): void {
```
Comment (fraszczakszymon) [inline src/ad-engine/services/viewability-counter.ts]: `private`?

Comment (dmnsobczak) [reply src/ad-engine/services/viewability-counter.ts]: Then the unit tests won't work :(

File: src/ad-engine/services/local-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmnsobczak) [reply src/ad-engine/services/local-storage.ts]: Done!

[the remark this answers is not in the document]
Comment (Bielik20) [reply src/ad-engine/services/local-storage.ts]: 💎

**Pull request merged**